### ISSUE TITLE
di.fm now unsupported with free account & still not worked without logging in  

### ISSUE THREAD

**Issue by Quqas** [lethak/digitally_imported_userscript#63]
Hi, a new trouble with di.fm  

If not logged in, music stop after 10-30 sec and redirect to registration page,

If log to free account - most channels is blocked to play,

In addition if logged to free account i receive geoblock, and only vpn help me (no geoblock if not logging in)

Any ideas to fix login needed?  

p.s. I understand that some functions are impossible without an account, like favorites, but at least basic and non-stop playback I think are possible

**Comment by lethak** [comment]
I am sad to hear the news, I can confirm its happening.
Looks like the same system used on non-DI sites for a while now.

![image](https://user-images.githubusercontent.com/1705300/225791249-0441d65f-928f-44e7-8963-a0dc040274d9.png)
![image](https://user-images.githubusercontent.com/1705300/225791276-ce622c48-890e-40d2-a86a-b330a5557bad.png)

As for the rest, I see no point on spending time on anonymous users when you can create a free account and much of the stuff only works with an account.

As for your geofencing problem, I am really curious what is that all about. Did they blocked users from Russia or something like that ? Have they announced it somewhere ?

- Have you tried a VPN in the same country as you ? 
- is it just your IP banned and not the whole country ?
- Did you downloaded too much too fast ?
- The same account works with outside VPN ?

If Russia is the problem, I can see why here (from twitter)
![image](https://user-images.githubusercontent.com/1705300/225802455-aeefaab0-2807-455e-8cf1-71405515600a.png)

And like I said on this subject before, this is on the backend, no workaround this on the client. I have taken a peak and found no trace in the current client code of any geoblocking feature.

**Comment by Quqas** [comment]
i tested many times.
--no account differences. only ip matter (all country. many providers)
--with vpn any account work fine. i.e. account not banned.

so I resigned myself to geoblock
interesting that without a login there are no restrictions

so it seems to me that if you redo the script, it's easier to remove the registration redirect after 30 seconds than to fully emulate a premium account and unblock channels (but if this is possible, then that's great)

**Comment by Quqas** [comment]
more info.

 different result from chrome incognito mode or normal mode **without login**.

in incognito all channel unblocked for play. in normal (even clear cache\cookie) still blocked.

without login i have only 30 sec playing from true ip and about 30 minute from vpn ip. in both case redirect to registration page

**Comment by lethak** [comment]
Unblocking channels is easy because I added this feature years ago for the other apps of AudioAddict (zenrradio, etc)

I already looked at removing the registration wall as an anonymous user and it was challenging enough that I did not wanted to spend countless hours to try to do this.  However, you are in luck, the next version of the script will be introducing this feature as I made a recent breakthrough.

**Comment by Quqas** [comment]
> Unblocking channels is easy because I added this feature years ago

I hope the new type of blocking will not cause problems

> I already looked at removing the registration wall as an anonymous user

May be this script helpful https://greasyfork.org/en/scripts/19993-ru-adlist-js-fixes/code

```
'di.fm': () => scriptLander(() => {
            let log = false;
            // wrap global app object to catch registration of specific modules
            let _di = win.di;
            Object.defineProperty(win, 'di', {
                get() {
                    return _di;
                },
                set(vl) {
                    if (vl === _di)
                        return;
                    if (log) _console.trace('di =', vl);
                    _di = new Proxy(vl, {
                        set(di, name, vl) {
                            if (vl === di[name])
                                return true;
                            if (name === 'app') {
                                if (log) _console.trace(`di.${name} =`, vl);
                                if (!('module' in vl))
                                    return;
                                vl.module = new Proxy(vl.module, {
                                    apply(module, that, args) {
                                        if (/Wall|Banner|Detect|WebplayerApp\.Ads/.test(args[0])) {
                                            let name = args[0];
                                            if (log) _console.log('wrap', name, 'module');
                                            if (typeof args[1] === 'function')
                                                args[1] = new Proxy(args[1], {
                                                    apply(fun, that, args) {
                                                        if (args[0]) // module object
                                                            args[0].start = () => _console.log('Skipped start of', name);
                                                        return Reflect.apply(fun, that, args);
                                                    }
                                                });
                                        } // else log && _console.log('loading module', args[0]);
                                        if (args[0] === 'Modals' && typeof args[1] === 'function') {
                                            if (log) _console.log('wrap', name, 'module');
                                            args[1] = new Proxy(args[1], {
                                                apply(fun, that, args) {
                                                    if ('commands' in args[1] && 'setHandlers' in args[1].commands &&
                                                        !Object.hasOwnProperty.call(args[1].commands, 'setHandlers')) {
                                                        let _commands = args[1].commands;
                                                        _commands.setHandlers = new Proxy(_commands.setHandlers, {
                                                            apply(fun, that, args) {
                                                                const noopFunc = name => () => _console.log('Skipped', name, 'window');
                                                                for (let name in args[0])
                                                                    if (name === 'modal:streaminterrupt' ||
                                                                        name === 'modal:midroll')
                                                                        args[0][name] = noopFunc(name);
                                                                delete _commands.setHandlers;
                                                                return Reflect.apply(fun, that, args);
                                                            }
                                                        });
                                                    }
                                                    return Reflect.apply(fun, that, args);
                                                }
                                            });
                                        }
                                        return Reflect.apply(module, that, args);
                                    }
                                });
                            }
                            di[name] = vl;
                        }
                    });
                }
            });
            // don't send errorception logs
            Object.defineProperty(win, 'onerror', {
                set(vl) {
                    if (log) _console.trace('Skipped global onerror callback:', vl);
                }
            });
        }),
```

years ago it's worked fine, but not now :(

**Comment by lethak** [comment]
I'll never complain about the maintainability of my code again :D joke aside, this is far more complex than just removing a modal now.

**Comment by lethak** [comment]
[v2023.3.54](https://github.com/lethak/LTKDIFMU-userscript/commit/5ebb7c2290f0db284a20fca46b25e009b7f043e8) should improve things. Clear localStorage and cookies after upgrading the script and reload page
